scaladoc-js/src/searchbar/SearchbarComponent.scala

```diff
@@ -32,26 +34,65 @@ class SearchbarComponent(val callback: (String) => List[PageEntry]):
       })
       wrapper
 
-  def handleNewQuery(query: String) =
-    val result = callback(query).map(_.toHTML)
+  def handleNewFluffQuery(matchers: List[Matchers]) =
+    val result = engine.query(matchers).map(_.toHTML(inkuire = false))
     resultsDiv.scrollTop = 0
     while (resultsDiv.hasChildNodes()) resultsDiv.removeChild(resultsDiv.lastChild)
     val fragment = document.createDocumentFragment()
     result.take(resultsChunkSize).foreach(fragment.appendChild)
     resultsDiv.appendChild(fragment)
     def loadMoreResults(result: List[raw.HTMLElement]): Unit = {
       resultsDiv.onscroll = (event: Event) => {
-          if (resultsDiv.scrollHeight - resultsDiv.scrollTop == resultsDiv.clientHeight)
-          {
-              val fragment = document.createDocumentFragment()
-              result.take(resultsChunkSize).foreach(fragment.appendChild)
-              resultsDiv.appendChild(fragment)
-              loadMoreResults(result.drop(resultsChunkSize))
-          }
+        if (resultsDiv.scrollHeight - resultsDiv.scrollTop == resultsDiv.clientHeight) {
+          val fragment = document.createDocumentFragment()
+          result.take(resultsChunkSize).foreach(fragment.appendChild)
+          resultsDiv.appendChild(fragment)
+          loadMoreResults(result.drop(resultsChunkSize))
+        }
       }
     }
     loadMoreResults(result.drop(resultsChunkSize))
 
+  extension (s: String)
+    def toHTMLError =
+      val wrapper = document.createElement("div").asInstanceOf[html.Div]
+      wrapper.classList.add("scaladoc-searchbar-result")
+      wrapper.classList.add("monospace")
+
+      val errorSpan = document.createElement("span").asInstanceOf[html.Span]
+      errorSpan.classList.add("search-error")
+      errorSpan.textContent = s
+
+      wrapper.appendChild(errorSpan)
+      wrapper
+
+  var timeoutHandle: SetTimeoutHandle = null
+  def handleNewQuery(query: String) =
+    clearTimeout(timeoutHandle)
+    resultsDiv.scrollTop = 0
+    resultsDiv.onscroll = (event: Event) => { }
+    while (resultsDiv.hasChildNodes()) resultsDiv.removeChild(resultsDiv.lastChild)
+    val fragment = document.createDocumentFragment()
+    parser.parse(query) match {
+      case EngineMatchersQuery(matchers) =>
+        handleNewFluffQuery(matchers)
+      case BySignature(signature) =>
+        timeoutHandle = setTimeout(1.second) {
+          val loading = document.createElement("div").asInstanceOf[html.Div]
+          loading.classList.add("loading-wrapper")
+          val animation = document.createElement("div").asInstanceOf[html.Div]
+          animation.classList.add("loading")
+          loading.appendChild(animation)
+          resultsDiv.appendChild(loading)
+          inkuireEngine.query(query) { (p: PageEntry) =>
+            resultsDiv.appendChild(p.toHTML(inkuire = true))
+          } { (s: String) =>
+            animation.classList.remove("loading")
+            resultsDiv.appendChild(s.toHTMLError)
+          }
+        }
+    }
```

Comment:
I feel like the current solution is missing another layer of engine. For me, there should be one composite engine which runs parser and delegates query to proper search engine. The benefits from that way would be: keeping SearchbarComponent logic easy (we could return to simple callback `String => List[PageEntry]` which would be implemented in composite engine) and separating frontend from backend(currently we call parser and engines in frontend part).

Reply:
Created #12995